Add IterableDataset

ssnl opened the pull request:
This is a modified version of https://github.com/pytorch/pytorch/pull/14705 since commit structure for that PR is quite messy.

1. Add `IterableDataset`.
3. So we have 2 data loader mods: `Iterable` and `Map`.

    1. `Iterable` if the `dataset` is an instance of `IterableDataset`
    2. `Map` o.w.

3. Add better support for non-batch loading (i.e., `batch_size=None` and `batch_sampler=None`). This is useful in doing things like bulk loading.
3. Refactor `DataLoaderIter` into two classes, `_SingleProcessDataLoaderIter` and `_MultiProcessingDataLoaderIter`. Rename some methods to be more generic, e.g., `get_batch` -> `get_data`.
4. Add `torch.utils.data.get_worker_info` which returns worker information in a worker proc (e.g., worker id, dataset obj copy, etc.) and can be used in `IterableDataset.__iter__` and `worker_init_fn` to do per-worker configuration.
5. Add `ChainDataset`, which is the analog of `ConcatDataset` for `IterableDataset`.
7. Import torch.utils.data in `torch/__init__.py`
9. data loader examples and documentations
10. Use `get_worker_info` to detect whether we are in a worker process in `default_collate`


Closes https://github.com/pytorch/pytorch/issues/17909, https://github.com/pytorch/pytorch/issues/18096, https://github.com/pytorch/pytorch/issues/19946, and some of https://github.com/pytorch/pytorch/issues/13023

ssnl committed:
Import torch.utils.data in torch/__init__.py

ssnl committed:
Add IterableDataset class and ChainDataset class

ssnl committed:
Separate DataLoaderIter into a single-process one and a multi-process one

Support three data loading modes in DataLoader, including IterableDataset

  Note that for IterableDataset, we need to consider scenario that
  a replica on exhausts before others do. Thus, the multiprocessing itera
  now maintains a worker status for each worker, and terminate when all
  workers are finished. In addition, in __next__, an outer loop is added
  for case when recv_idx maps to a task that triggers IterationEnd at one
  worker, so we can properly bump recv_idx to the next valid one.

Add torch.utils.data.get_worker_info for configuring replicas differently

Detect and report error in worker init

ssnl committed:
Add test for IterableDataset loading, worker init detection, get_worker_info

ssnl committed:
Update memory pinning doc; add reference so data loader doc can reference it

ssnl committed:
Comprehensive data loader doc

SsnL commented:
this is currently awaiting (1) #19421 and (2) filling in the docs

pritamdamania87 commented on torch/utils/data/dataloader.py:
```diff
@@ -21,29 +24,52 @@
+        collate_fn (callable, optional): merges a list of samples to form a
+            mini-batch of Tensor(s).  Used when using batched loading from a
+            map-style dataset.
+        convert_fn (callable, optional): converts a sample to Tensor(s).  Used
```
Could you elaborate a bit in terms of how and when collate_fn and convert_fn are used? I'm a little confused in terms of how the two differ.

SsnL replied:
`collate_fn` is used when batch loading (i.e., the current supported kind), to collate a list of samples into a batched sample, e.g., the two element batch `[(A1, A2), (B1, B2)]` into `torch.as_tensor([A1, B1]), torch.as_tensor([A2, B2])`.

`convert_fn` just does conversion, so if a sample is `[(A, B), (C, D)]` it can just call `as_tensor` on each.

pritamdamania87 replied:
I see, could you also elaborate a bit in the comments as well so the difference is clear?

SsnL replied:
Yes, I will do it in the doc.

pritamdamania87 commented on torch/utils/data/_utils/signal_handling.py:
```diff
@@ -32,9 +32,12 @@
-from torch._C import _set_worker_pids, _remove_worker_pids, _error_if_any_worker_fails, _set_worker_signal_handlers  # noqa: F401
 from . import IS_WINDOWS
 
+# Some of the following imported functions are not used in this file, but are to
```
Should we import these in those files instead then?

SsnL replied:
no, they are supposed the be exposed under the meaningful namespace `signal_handling.XXX`.

pritamdamania87 commented on torch/utils/data/_utils/worker.py:
```diff
@@ -75,8 +125,37 @@ def _worker_loop(dataset, index_queue, data_queue, done_event, collate_fn, seed,
+            if mode == _DataLoaderStrategy.Iterable:
+                dataset_iter = iter(dataset)
+
+        except Exception:
```
Does this try, except just need to be around init_fn?

SsnL replied:
And `iter`

pritamdamania87 commented on docs/source/data.rst:
```diff
@@ -2,11 +2,219 @@ torch.utils.data
+The behavior of this mode is roughly equivalent with::
+
+    for index in sampler:
+        yield convert_fn(dataset[index])
```
Why do we use a `convert_fn` here, as per the documentation, shouldn't the dataloader just return what it got from the dataset?

SsnL replied:
No, `collate_fn` does automatic conversion from a bunch of things to tensors. `convert_fn` should do the same.

pritamdamania87 commented on torch/utils/data/dataloader.py:
```diff
@@ -619,11 +769,13 @@ def _shutdown_workers(self):
-                    # Indicate that no more data will be put on this queue by the
-                    # current process.
-                    q.close()
+                for i, q in enumerate(self.index_queues):
```
Can we reuse `shutdown_worker` here?

SsnL replied:
sure

pritamdamania87 commented on torch/utils/data/dataset.py:
```diff
@@ -91,8 +190,36 @@ def cummulative_sizes(self):
+        for d in self.datasets:
+            assert isinstance(d, IterableDataset), "ChainDataset only supports IterableDataset"
+            for x in d:
+                yield x
```
How would this chaining work? Don't we call iter() only once at the beginning of each worker loop?

SsnL replied:
Could you elaborate why this would not work? This returns a generator, which is the iterable.

pritamdamania87 replied:
I haven't used generators, so I might be missing something here. When you call `dataset_iter = iter(dataset)` in the worker_loop and the dataset is a ChainedDataset, isn't dataset_iter an iterator over a chain of IterableDataset and not really an iterator over the data itself?

SsnL replied:
This goes through each `IterbleDataset` and yields each element from it. So it is yielding the data, not the dataset objects.

SsnL replied:
This probably helps for understanding: https://docs.python.org/3/library/itertools.html#itertools.chain

pritamdamania87 replied:
Oh my bad, for some reason I misread and thought it was yielding each dataset. Nevermind, this makes sense now.

SsnL commented on docs/source/data.rst:
```diff
@@ -2,11 +2,219 @@ torch.utils.data
+Randomness in Multiprocessing Data Loading
+""""""""""""""""""""""""""""""""""""""""""
+
+
```
doc isn't finished yet!

SsnL commented on torch/utils/data/_utils/__init__.py:
```diff
@@ -58,4 +59,35 @@ def _set_python_exit_flag():
+from .. import Sampler
+
+
+class ConstantSampler(Sampler):
```
Good point! Will kill it.

ssnl committed:
kill _use_shared_memory and make test_get_worker_info error more descriptive

SsnL commented:
@pritamdamania87 doc is done!

ssnl committed:
lint

pritamdamania87 commented:
@apaszke Soumith mentioned it'd be nice to have you review parts of this PR. I've looked through some of the new functionality added in this PR (IterableDataset), but it would be good if you can take a look at some of the changes to the existing code structure. Thanks!

ssnl committed:
private-fy some multiprocessing iter methods; refactor a test

SsnL commented:
@pritamdamania87 do let me know if this breaks anything internal

pritamdamania87 commented:
> @pritamdamania87 do let me know if this breaks anything internal

Looks like renaming pin_memory_batch breaks a bunch of stuff which does `from torch.utils.data._utils.pin_memory import pin_memory_batch`. Also, it looks like I see a lot of these errors coming from child processes: `TypeError: unhashable type: 'list'`. Not sure what might be the source of those though.

SsnL commented:
@pritamdamania87 for the first error, I prefer fixing those who import private helpers. For the second, it’d be great if you can show/message me the trace, or at least the non confidential parts. Thanks!

pritamdamania87 commented on torch/utils/data/dataloader.py:
```diff
@@ -55,131 +81,84 @@ class DataLoader(object):
+                        sampler = SequentialSampler(dataset)
+                batch_sampler = BatchSampler(sampler, batch_size, drop_last)
+
+            self.sampler = batch_sampler
```
Looks like the `TypeError: unhashable type: 'list'` error stems from this change. Code expects the sampler to return List[int], but it actually returns List[List[int]] since sampler is initialized to BatchSampler whereas earlier it would be initialized to RandomSampler.

SsnL replied:
oh i see. okay, i'll fix the type hints as well!

SsnL replied:
hmm wait. `.sampler` isn't even typed in `dataloader.pyi`. Are some call sites just accessing nonpublic members and enforcing types on them?.... Those internal code should be fixed as well.

pritamdamania87 replied:
I agree those should be fixed, although we can't change the behavior to initialize sampler to BatchSampler since the code relies on the behavior of RandomSampler (not the type).

pritamdamania87 replied:
Can we fix the default behavior here?

SsnL replied:
I don't know why this needs fixing and I don't quite understand your argument about `RandomSampler` and `BatchSampler`. This does not change the data loader's behavior whatsoever. 

IIUC, this is an issue where some downstream code uses these private attributes and expects them to behave whatever they expect it to be. For such cases, I strongly don't agree that we need to fix. If they hack our library and decide to have test relying on the hack, they should be the one fixing. I have suffered from such things already before when doing the dataloader refactoring.

These said, I'm now not at facebook so I can't yell at them. The new commit should hopefully "unbreak" their broken test.

SsnL replied:
It would be cool if you can tell me if this does unbreak the test.

SsnL replied:
Any updates? @pritamdamania87

pritamdamania87 replied:
Sorry I missed this message earlier. Yes, the latest changes do fix the tests. Thanks a lot!

apaszke commented on docs/source/data.rst:
```diff
@@ -2,11 +2,342 @@ torch.utils.data
+Map-style datasets
+^^^^^^^^^^^^^^^^^^
+
+A map-style dataset is one that implements the ``__getitem__`` protocol,
```
It would be nice to mention it as `__len__` + `__getitem__` because you really need both of those.

apaszke commented on docs/source/data.rst:
```diff
@@ -2,11 +2,342 @@ torch.utils.data
+arguments, a :class:`~torch.utils.data.DataLoader` operates in one of three
+following strategies:
+
+* `Batched loading from a map-style dataset (default)`_
```
It's not the default if you pass in an interable dataset, so I don't think we should mention that, unless you divide those points into two categories.

apaszke commented on docs/source/data.rst:
```diff
@@ -2,11 +2,342 @@ torch.utils.data
+  behavior.
+
+  With the default arguments, a :class:`~torch.utils.data.DataLoader`
+  loads data as batches of size ``1`` with indices sampled without replacement.
```
nit: _sampled without replacement_ assumes `shuffle=True`

apaszke commented on docs/source/data.rst:
```diff
@@ -2,11 +2,342 @@ torch.utils.data
+Working with :attr:`collate_fn`
+"""""""""""""""""""""""""""""""
+
+For instance, if each data sample consists of a 3-channeled image and an
```
nit: shouldn't this simply be _3-channel image_?

apaszke commented on docs/source/data.rst:
```diff
@@ -2,11 +2,342 @@ torch.utils.data
+  for ``list`` s, ``tuple`` s, ``namedtuple`` s, etc.
+
+Users may use customized :attr:`collate_fn` to achieve custom batching, e.g.,
+along a dimension other than the first.
```
_or to add support for custom data types_

apaszke commented on docs/source/data.rst:
```diff
@@ -2,11 +2,342 @@ torch.utils.data
+Loading individual members of a map-style dataset
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Sometimes it is cheaper to directly load batched data (e.g., bulk reads from a
```
This is a few feature of this PR, right? I'm not 100% convinced that we want to implement it this way. What you do here is that you really have a `BatchDataset`, which takes an index of a batch to return. This has a few problems like the fact that the division into batches does not depend on the sampler, and you need to reset that at every epoch too. Hence, I think that in this mode we should really call `__getitem__` with a list or a tuple of indices to return, allowing the dataset to perform a batched query.

apaszke replied:
Also, then we can make the dataset return a list/tuple of answers to that query, and we don't need to add `convert_fn`, because it still corresponds to `collate_fn`

SsnL replied:
Yes. You are right about what this mode does. I should word this part of the documentation better because I think the mode supports very similar to what you describes. Let me describe the use case I had in mind when implementing it, which I believe will make things clearer:

Users pass in a `sampler` that each time yields batched indices as a list (or as any type really, e.g., partition idx of a dataset in a database), and a `dataset` object whose `__getitem__` can handle such batch indices type, and will properly do bulk loading, and yield batched tensors. They also have the choice to either do batching themselves in `__getitem__` if their bulk loading returns data in nice contiguous chunks, or supply `default_collate` as the `convert_fn`.

Let me know if this makes sense to you. I'd be happy to hear your suggestions and opinions on it!

Finally, I want to note that `convert_fn` is still needed even if we kill / completely change this mode's behavior, as it is needed for the `IterableDataset` case.

apaszke commented on docs/source/data.rst:
```diff
@@ -2,11 +2,342 @@ torch.utils.data
+
+For multi-process loading, each worker process gets a different copy of
+:attr:`dataset`, so it is often desired to configure each copy independently to
+avoid having duplicate data returned from the workers. See the documentations of
```
nit: see the documentation (without s)?

apaszke commented on torch/utils/data/_utils/pin_memory.py:
```diff
@@ -36,27 +36,27 @@ def _pin_memory_loop(in_queue, out_queue, device_id, done_event):
-        return [pin_memory_batch(sample) for sample in batch]
-    elif hasattr(batch, "pin_memory"):
-        return batch.pin_memory()
+def pin_memory_data(data):
```
Or just `pin_memory`? 😉

apaszke commented on torch/utils/data/_utils/worker.py:
```diff
@@ -75,8 +123,37 @@ def _worker_loop(dataset, index_queue, data_queue, done_event, collate_fn, seed,
+        # When such things happen, an `IterableDatasetStopIteration` object is
+        # sent over to the main process with the ID of this worker, so that the
+        # main process won't send more tasks to this worker, and will send
+        # `None` to this worker to properly exit it.
```
Why do we even bother with sending anything from master to workers in iterable mode? Because you want to bound the number of batches awaiting in memory? Couldn't we simply use an atomic IPC counter for that purpose?

SsnL replied:
> Because you want to bound the number of batches awaiting in memory?

Yep, that is the main reason. I do agree that an simple atomic IPC would probably suffice the majority of the work. But it would require adding more than just that if we do not use `queue`s at all. Consider the following requirements, all of which we spent significant amount of work to make sure work with `queue` communication:

1. We want data loading order to be deterministic (conditioned on seeds). 

    Following the current DataLoader design, we index the workloads to properly track which is the next batch to return. This allows the worker to receive an `index` and send it back. Admittedly this could also be done if we just for each worker keep a FIFO queue of indices of all work loads, but the existing code does so naturally with `queues`. 

3. Signaling the workers to quit. 

    Currently we use sending `None` to workers as the authority signal to terminate. Yes, we could use another `mp.Event` to do so (`done_event` is not exactly for this purpose), but the existing code does so with `queue`s in a quite robust way.

2. Do not hang.

   We took sooo long to build all these mechanisms for `queue` communication (hopefully) correctly to prevent hanging. Implementing new ways to signal workers using semaphores or `mp.Condition`s could be tricky (although I admit I didn't spend too much time thinking about the details).

Reinventing these wheels would be a lot of work and might be hard to get right. So I just went with the `queue`s, which also nicely reduce code duplication and are probably implemented under the hood as semaphores and pipes anyways.

apaszke replied:
Ok, this makes sense. I just wanted to clarify that. Thanks for the detailed respose!

apaszke commented on torch/utils/data/dataloader.py:
```diff
@@ -21,29 +24,52 @@
+
+    * batched loading from a map-style dataset (default)
+
+    * Loading individual members of a map-style dataset
```
nit: make this lowercase to be consistent

apaszke commented on torch/utils/data/dataloader.py:
```diff
@@ -55,131 +81,84 @@ class DataLoader(object):
+        self.convert_fn = convert_fn
+        self.collate_fn = collate_fn
+        if isinstance(dataset, IterableDataset):
+            # FIXME: check default for some args
```
Please do! At least check that the sampler is `None`. We should do this before we land this PR, otherwise it's pretty bad UX

apaszke commented on torch/utils/data/dataloader.py:
```diff
@@ -190,13 +169,91 @@ def __setattr__(self, attr, val):
+        return len(self.sampler)
+
+
+class _BaseDataLoaderIter(object):
```
This class has so many `if`s for the iterable dataset, that I'm not sure if it shouldn't be a separate class. If you don't want to repeat all the assignments in `__init__` make them have a superclass, and then override the behavior in subclasses.

SsnL replied:
Yes, I do agree that there is a lot of `if`s here, making it kind of ugly. 

However, I disagree that this should be made into different subclasses. The class inheritance structure is made so that most of the `mode`-dependent code lives in these functions implemented by `_BaseDataLoaderIter`, while `_SingleProcessDataLoaderIter` and `_MultiProcessDataLoaderIter` mostly does mode independent things. Having mode-dependent `DataLoaderIter` classes will be a lot of subclasses, and hurts readability and cleanness IMO.

apaszke replied:
I'm just afraid that right now it is _highly unclear_ which attributes are defined when, and so checking the correctness of the code below if extremely hard. I agree that having 6 subclasses is a bit much, but maybe we can factor out the serial/parallel and iterable/batched/map components and then compose them separately? Right now we lack the second split.

SsnL replied:
I see your point. I'll give it a try to refactor.

SsnL replied:
I did a refactoring where I applied a dummy sampler for `IterableDataset` (which is needed to support batching for its case), and moved the shared logic of actually fetching from dataset to a separate class. Now the `if`-statements on the dataset type is minimal in `_BaseDataLoaderIter` and its subclasses: `1` in total.

apaszke reviewed (commented):
Looks pretty good, but in the future please split PRs like this into parts. A few of the points you list are easily separable from others, but make this diff unnecessarily large and hard to review. For that reason I didn't follow the implementation of `_MultiProcessingDataLoaderIter `.

I have a single followup question: do we support batching of iterable datasets? It seems like a reasonable thing to do, but I'm not convinced that it works right now.

SsnL commented:
@apaszke Thanks for the review!  I will address most of the points you made in new updates. I also replied to some of the comments above. Let me what you think!

Also, I do agree that this PR is unncessarily big. The commit structure in the old PR got messed up so much that I couldn't easily break things apart. I apologize and will make sure to properly divide up PRs in future :)

pritamdamania87 commented on torch/utils/data/dataloader.py:
```diff
@@ -425,80 +491,77 @@ class _DataLoaderIter(object):
+
+        self.worker_init_fn = loader.worker_init_fn
+        self.worker_queue_idx_cycle = itertools.cycle(range(self.num_workers))
+        self.worker_result_queue = multiprocessing.Queue()
```
I'm wondering if this queue needs to have some max capacity? It seems like we might run into memory issues if we have a very large dataset and the workers produce work at a much faster than what the application can consume.

SsnL replied:
there is an upper bound of number of outstanding jobs (2*num_workers), so no this does not need a max capacity.

SsnL replied:
This is also not a new behavior introduced by this pr.

pritamdamania87 commented on torch/utils/data/dataloader.py:
```diff
@@ -55,131 +81,84 @@ class DataLoader(object):
-              each worker in :attr:`worker_init_fn`, and use it to set other
-              seeds before data loading.
-
     .. warning:: If ``spawn`` start method is used, :attr:`worker_init_fn` cannot be an
```
Had another question regarding existing code. This comment seems to suggest we can specify a start method for multiprocessing. Unless I'm missing something it doesn't seem like the DataLoader interface supports a start method.

SsnL replied:
The mp package already provides a way, i.e., `mp.set_start_method`. It sets the start method globally, so DataLoader shouldn't do that automatically. That said, we could let the data loader to take in a mp context, in case that people want to easily use different start methods in the same program.

pritamdamania87 replied:
Another option would be to accept a start_method in the dataloader and build an mp context inside the dataloader and set the start method appropriately.

SsnL replied:
Sure sure, we can even let it take in both, but that is really unrelated with this pr.

pritamdamania87 replied:
Yes, I completely understand, I just thought I'd mention this for future enhancements.

ssnl committed:
address comments

apaszke commented:
I'm still not convinced that we need `convert_fn`. Is it just me, or is there no way to get the data loader to collate batches out of an iterable dataset that returns individual examples? I really feel like we should support the case of `DataLoader(iterable_dataset, batch_size=10)`. Then, the `collate_fn` we use today fits perfectly with the new code, and the case of batched datasets can still use the same argument (which will operate on a single element instead).

dzhulgakov reviewed (commented):
It'd be nice to land this PR soon as we have some stuff that depends on it.

Regarding @apaszke's comment, I think the convert_fn is needed to avoid prepending an extra dimension to tensors (which `collate_fn([data])` would do)l.

As for additional rebatching - DataLoader(iterable_dataset, batch_size) should work, but stuff would be converted to tensors first and then batched

SsnL commented:
@apaszke Okay, we can support `DataLoader(iterable_dataset, batch_size=10)`, which uses `collate_fn`. But I don't think that it can replace `convert_fn`. There are still many cases where people want to do bulk loading in either `__getitem__` or `__next__`, where with `collate_fn` the returned bulk loaded result will be put in a list, and one needs to pass as `collate_fn` the `lambda x: collate_fn(x[0])` .

@dzhulgakov Yes, that is indeed the motivation for `convert_fn`.

SsnL commented:
@dzhulgakov Yes, I hope to get to this really soon. Now is between semester ending and NeurIPS deadline so things get quite a bit crazy. Sorry for the delay. I'll push an update today that addresses most of the comments.

pritamdamania87 commented on test/test_dataloader.py:
```diff
@@ -505,6 +661,15 @@ def _test_error(self, loader):
                                  math.ceil(float(len(loader.dataset)) / loader.batch_size))
                 return
 
+    def test_error_in_init(self):
```
Looks like `test_error_in_init` is defined twice.

pritamdamania87 commented on test/test_dataloader.py:
```diff
@@ -459,6 +604,17 @@ def error_worker_init_fn(_):
+
+
+# used with test_error_in_init
+def error_worker_init_fn(_):
```
`def error_worker_init_fn` defined twice as well.

ssnl committed:
Address various review comments
Make batched vs. non-batched a separate dimension than map vs iterable, so now IterableDataset can do batched loading.

SsnL commented:
@apaszke @pritamdamania87 @dzhulgakov I've updated the PR substantially to isolate batching from dataset style, i.e., now you can use batching with an `IterableDataset`. 

Now there are basically three independent axes one can tune:
+ `iterable` vs `map`
+ `batch` vs `non-batch`
+ `single-` vs `multi-` process

Doc is not yet updated since it will be a lot of work to rewrite many sections, but I hope to get to it within a day or two.

ssnl committed:
doc update

apaszke commented on torch/utils/data/dataloader.py:
```diff
@@ -426,80 +542,77 @@ class _DataLoaderIter(object):
+        self.workers = []
+        # A list of booleans representing whether each worker has exhausted
+        # its iterable dataset object. Always contain all Trues if not using
+        # an iterable dataset (i.e., if mode != Iterable).
```
This comment is highly misleading, because this list actually holds a negation of this statement for each worker.

apaszke commented on torch/utils/data/dataloader.py:
```diff
@@ -551,73 +669,115 @@ def _get_batch(self):
+                data = self.task_info.pop(self.rcvd_idx)[1]
+                return self._process_data(data)
+
+            assert (not self.shutdown and self.tasks_outstanding > 0)
```
please please please never use parenthesis with an `assert`

apaszke reviewed (commented):
The code really looks good at this point, but there are a few minor surprising design decisions that I'd like us to change or discuss. Apart from that docs need to be rewritten to stop mentioning `convert_fn`.

EvenOldridge commented:
Is https://github.com/pytorch/pytorch/blob/1834c27184387175b39722b750c153a37754766c/torch/utils/data/_utils/fetch.py#L43 inverting the logic of is_batched?  The access pattern of the batched vs non-batched seem reversed to me.

ssnl committed:
Disallow batch_sampler with IterableDataset
default_convert now only touches np arrays
Rename is_batched to auto_collation for clarity
Fix default __len__ in some abstract classes
Improve various comments

SsnL commented on torch/utils/data/_utils/collate.py:
```diff
@@ -9,62 +9,83 @@
+    elif isinstance(data, int_classes):
+        return torch.tensor(data, dtype=torch.long)
+    elif isinstance(data, float):
+        return torch.tensor(data, dtype=torch.double)
```
I updated so that `default_convert` only touches convertable np arrays.

SsnL commented on torch/utils/data/dataloader.py:
```diff
@@ -55,132 +98,123 @@ class DataLoader(object):
+                raise RuntimeError(
+                    "DataLoader with IterableDataset: expected unspecified "
+                    "sampler option, but got sampler=True")
+            elif batch_sampler is not None and num_workers > 0:
```
fair point. disabled in new commit.

ssnl committed:
Add & improve a bunch of arg checks

SsnL commented on docs/source/data.rst:
```diff
@@ -2,11 +2,354 @@ torch.utils.data
+In this mode, each time an iterator of a :class:`~torch.utils.data.DataLoader`
+is created (e.g., when you call ``enumerate(dataloader, 0)``), :attr:`num_workers`
+worker processes are created. At this point, the :attr:`dataset`, :attr:`collate_fn`,
+:attr:`convert_fn` and :attr:`worker_init_fn` are passed to each
```
fixed!

thiagocrepaldi commented:
@SsnL Great job, I have a change on my private branch that does basically the same thing, but the documentation is not as nicer as yours and you added a couple more features :)

Is there any ETA on when this is going to be merged? I would like to build on top of this change before the release 1.2 comes up in July

soumith reviewed (approved):
adam said shipit

SsnL commented:
Hey @thiagocrepaldi , thanks! I'm embarrassed that this is dragged for so long and that I couldn't address review comments as fast as I hoped.  :)

Okay halfway through writing this comment, @soumith stamped this lol.

SsnL commented:
@soumith let's wait for this CI turnaround.

ssnl committed:
fix flakiness caused by cancel_join_thread

apaszke commented on torch/utils/data/dataloader.py:
```diff
@@ -55,132 +89,138 @@ class DataLoader(object):
+                 :ref:`multiprocessing-best-practices` on more details related
+                 to multiprocessing in PyTorch.
+
+    .. note:: ``len(dataloader)`` is determined by length of the sampler used.
```
This is definitely too low-level. The users should not care about the particular implementation we use. We should say something along the lines of "`__len__` is a heuristic based on the length of the sampler. Note that subclasses of `IterableDataset` always use an infinite sampler, so one should not query this method unless they work with a map-style dataset." This way we can avoid all the problems related to multiprocess sampling from iterable datasets and make it easy to understand.

We should also make it an error to use `__len__` with a non-map dataset.

SsnL replied:
fixed!

apaszke commented on torch/utils/data/dataloader.py:
```diff
@@ -426,80 +592,80 @@ class _DataLoaderIter(object):
+
+        self.index_queues = []
+        self.workers = []
+        # A list of booleans representing whether each worker still has word to
```
nit: word -> work

SsnL replied:
fixed!

apaszke reviewed (approved):
I think we should make `DataLoader.__len__` even more strict as it is today, and simplify the docs. Still, this PR is a beast, so I don't want to block it over that. We can update that in a follow up patch.

This looks good, thanks a lot for doing all this @SsnL!

ssnl committed:
address comments

SsnL commented:
Thanks @apaszke @soumith @pritamdamania87 !

ssnl committed:
say more about custom collate_fn in docs

ssnl committed:
Raise more meaning errors in len(loader) for iterable dataset, and test

bethunebtj commented:
Seems like the PR is finished? Good job! 
Btw, when will this be released? I'm looking forward to this. Again, thanks a lot!

dzhulgakov commented:
@bethunebtj - we're targeting 1.2 release at the end of July / early August hopefully. But you can always use nightlies :)

rfalcon100 commented:
Is this functionality available in nightly?

soumith commented:
@rfalcon100 yes

kuraga commented:
Dear Developers,
take a look at #120139, please.

---

Also, I'm very worried about

https://github.com/pytorch/pytorch/blob/v2.2.1/torch/utils/data/sampler.py#L70-L95

So,

```
    #   + `return NotImplemented`:
    #     Calling `len(subclass_instance)` raises:
    #       TypeError: 'NotImplementedType' object cannot be interpreted as an integer
```
But what should return `subclass_instance` without an implemented length value?
And what does it return if `class_instance.__len__` is not defined?

```
    #   + `raise NotImplementedError()`:
    #     This prevents triggering some fallback behavior. E.g., the built-in
    #     `list(X)` tries to call `len(X)` first, and executes a different code
    #     path if the method is not found or `NotImplemented` is returned, while
    #     raising a `NotImplementedError` will propagate and make the call fail
    #     where it could have used `__iter__` to complete the call.
```
Which `list(X)` tries to call `len(X)`?
```python
class A:
    def __iter__(self):
        for i in range(8):
            yield i
a = A()
print(list(a))
len(a)
```
```
[0, 1, 2, 3, 4, 5, 6, 7]
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[22], line 7
      5 a = A()
      6 print(list(a))
----> 7 len(a)

TypeError: object of type 'A' has no len()
```
Would `len(X)` have not been called if `__len__` had not been defined?

Thanks!
